### PR TITLE
Add AdminNavigationProvider to cleanup AdminMenu

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Instead of the `NavigationHelper.IsAdminMenu` helper that was introduced in this version, `AdminNavigationProvider` allows us to cleanup our code and build the navigation with less and cleaner code.

Since we already introduced a breaking change in 2.0.0 for `INavigationProvider`, I think this PR should be part of 2.0.0 this way people can modify their Admin Menu once and address all the improvements at once.

### CODE REVIEW

**Comment by hishamco** [comment]
Can we hold this after `2.0.0` we don't want to introduce any bugs accidentally, every PR in this stage is critical

**Comment by MikeAlhayek** [comment]
I am hoping for 2.0.0 because we did lots of enhancements to the AdminMenu including introducing the helper IsAdmin Helper which in this case was removed.

**Comment by hishamco** [comment]
I understand but I have many things to do before `2.0.0`, but last week you are telling that we needed to publish the release on TUE or THUR for that I deferred everything after even the PO Extractor Tool update

I suggest organizing the release period every time giving some time for test and integrating the newly added functionality and rejecting any modification in the preview unless it's critical. FYI I already saw such thing in many OSS including ASP.NET Core itself

**Comment by MikeAlhayek** [comment]
>  I deferred everything after even the PO Extractor Tool update

Nothing related to translation was changed here. So it should not impact anything you have done.

But I understand your concern. My concern is to reduce the amount of work needed that is going to be needed during upgrade. In this case, I think this PR is fair and will make things easy. Maybe instead of debating, checkout the PR and test it out?

File: src/OrchardCore/OrchardCore.Navigation.Core/NavigationProviderBase.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Navigation.Core/NavigationProviderBase.cs:35]: The only advantage of `Build` instead of `BuildAsync` is that you can return nothing instead of `ValueTask.CompletedTask`.

With only adding a filter on `name` I don't think this PR is a good thing since it's adding new concepts to discover/understand (NavigationProviderBase vs AdminNavigationProvider) where before anything would implement INavigationProvider and check the name.

Also adding a breaking change (IsAdminMenu) though this can be mitigated with an [Obsolete].

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/NavigationProviderBase.cs:35]: @sebastienros 

> Also adding a breaking change (IsAdminMenu) though this can be mitigated with an [Obsolete].

This isn't a breaking change because `IsAdminMenu` was added in 2.0 preview. So removing it now is not a breaking change. But if we don't include this PR in 2.0, then it will become a breaking change unless we use `[Obsolete]` to not make it breaking.

The main key here is to ask the users to make all the changes to the AdminMenu at once.

> With only adding a filter on name I don't think this PR is a good thing since it'

People don't have to follow our clean code approach and use this new base class, they can continue to write a longer code if they want to. With this approach it is just a much cleaner way to implement an navigation in out admin menu.

**Comment by hishamco** [comment]
> Maybe instead of debating, checkout the PR and test it out?

To be honest I have many things in PO Extractor Tool & Orchard Core Contrib. I told you PRs in the release time need time for testing & integration

**Comment by sebastienros** [comment]
OK with me on these conditions, and it's fine if you can find someone else to OK it without my conditions, you pick which approval you need ;)
- Rename `NavigationProviderBase` to `NamedNavigationProvider` such that it only adds the verification of the name, and there is a single async method to implement `protected virtual ValueTask BuildNavigationAsync(NavigationBuilder builder)`. This class also has a single constructor with a `string name` such that concrete implementations have to call into `base(name)`. The `Name` property is `protected { get; }` only.
- Rename `AdminNavigationProvider` to `AdminMenuNavigationProvider`, inheriting from `NamedNavigationProvider`.

**Comment by hishamco** [comment]
I already something quite similar to Seb suggestion in Orchard Core Contrib long time ago, I will check the PR very soon, but we need to fix the build issue first

File: src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs:3]: `AdminNavigationProvider` instead

File: src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs:3]: IMHO no need to introduce a new type, `AdminNavigationProvider` is enough. We can have `Name` as an overridable property, so if anyone wants to use a other name than `Admin` he could

File: src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs:3]: @hishamco It was `AdminNavigationProvider` but Seb requested to change it to `AdminMenuNavigationProvider` so we'll keep it as`AdminMenuNavigationProvider`

File: src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs:3]: `NamedNavigationProvider` is much cleaner where any one that want to create a base class can.

File: src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs:3]: I disagree with Seb :)

File: src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs:3]: You can check the Orchard 1, no need to add more base classes IMHO

File: src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs:3]: @sebastienros I am also okay with the original `AdminNavigationProvider`. Are you?

File: src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs:3]: I don't see anything wrong with the `NamedNavigationProvider` in this case.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs:3]: > public abstract class AdminMenuNavigationProvider : NamedNavigationProvider

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/NamedNavigationProvider.cs:3]: I like to keep the below: Also Seb is also okay with it. 

```
AdminNavigationProvider : NamedNavigationProvider
```

File: src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Navigation.Core/AdminMenuNavigationProvider.cs]: @hishamco I renamed this.

**Comment by hishamco** [comment]
Why the functional tests are skipped?

**Comment by MikeAlhayek** [comment]
@hishamco they are triggered after approval.

**Review by hishamco — APPROVED** [review]

**Comment by sebastienros** [comment]
> they are triggered after approval.

I like this 💯

**Pull request merged**
